kvm: volumes support

ppalucki opened the pull request:
because kvm flavor doesn't use systemd-nspwan, we cannot utilize
its feature to bind mount demanded volumes (both kinds), so this
commit moves that responsibility to systemd (idea thx to xnox)

how it works - check the [mount.go "package" docs](https://github.com/intelsdi-x/rkt/commit/c3b08193f17b7b7c093c84986f0d6bebe2495739#diff-5e4f009487e0dd19a243aaca7de5862bR18) for idea or commit message 

**TODOs**:
- [x] code reorganization (kvm subpackge - take #1219 into account)
- [ ] add rkt functional tests (**but first need to deal with [failures](https://github.com/intelsdi-x/rkt/issues?q=is%3Aopen+is%3Aissue+label%3Atests)**)
- [x] ACE validator says "mount point problem": `[    0.269975] ace-validator[85]: 2015/07/21 20:30:16 WARN: "/db" is not a mount point (againts "/") fsid are equal: syscall.Fsid{X__val:[2]int32{-926896844, 1387030233}} == syscall.Fsid{X__val:[2]int32{-926896844, 1387030233}}
`  closing because not related to kvm ( iaguis's https://github.com/coreos/rkt/pull/1176#issuecomment-127546032 )
- [x] investigate systemd templates as mount units (impossible because mount unit type is marked with ['no_instances' flag](https://github.com/systemd/systemd/blob/master/src/core/mount.c#L1846) and got errors from [here](https://github.com/systemd/systemd/blob/bd37a92297a89f00d03757cf860dff9a8ae59b16/src/core/unit.c#L181)
- [x] path creation move to path.go (invalid I don't want to mess `path.go` with additional 7 functions to build all what/where host/app combination folder names, which logic is strictly kvm flavor specific)
- [x] escaping of name of volumes, because are used for directory names (vol name is an ACName type so is valid as 9p tag or folder name)
- [x] ACE validator "AC_APP_NAME not proper" 
`[    0.263976] ace-validator[84]: ==> AC_APP_NAME not set appropriately (need "ace-validator-main", got "coreos.com/ace-validator-main")` (#1199)

ppalucki commented:
marked as WIP, because after last rebase and fixing appId to string coercion, validator doesn't work anymore

```sh
~/.../github.com/coreos/rkt (volumes-ac)$ sudo ./bin/rkt --insecure-skip-verify run --mds-register=false --volume=database,kind=empty --private-net ../ace-validator-main.aci ../ace-validator-sidekick.aci
2015/07/24 15:21:26 Preparing stage1


2015/07/24 15:21:44 Writing image manifest
2015/07/24 15:21:44 Loading image sha512-822af9f297d3c85126983907789980c16cccc6c336f88fc3cd00cbea9a8c7f93
2015/07/24 15:21:44 Loading image sha512-8d5966cac9fb72a07d6b958551bb9bcfb87d1dde26332037a4ac539a509e1a01
2015/07/24 15:21:44 Writing pod manifest
2015/07/24 15:21:44 Setting up stage1
2015/07/24 15:21:44 Wrote filesystem to /var/lib/rkt/pods/run/c6f15a6f-d37b-468b-9ef1-902ea984ec74
2015/07/24 15:21:44 Pivoting to filesystem /var/lib/rkt/pods/run/c6f15a6f-d37b-468b-9ef1-902ea984ec74
2015/07/24 15:21:44 Execing /init
Continuing with per-app isolators disabled: could not get own cgroup path: controller "name=systemd" not found
  # lkvm run -k stage1/rootfs/bzImage -m 128 -c 1 --name rkt-c6f15a6f-d37b-468b-9ef1-902ea984ec74
[    0.000000] Initializing cgroup subsys cpu
[    0.000000] Initializing cgroup subsys cpuacct
[    0.000000] Linux version 4.1.3rkt-v1 (ppalucki@gklab-154-092) (gcc version 4.8.4 (Ubuntu 4.8.4-2ubuntu1~14.04) ) #1 SMP Fri Jul 24 15:21:11 CEST 2015
[    0.000000] Command line: noapic noacpi pci=conf1 reboot=k panic=1 i8042.direct=1 i8042.dumbkbd=1 i8042.nopnp=1 console=ttyS0 earlyprintk=serial i8042.noaux=1 console=hvc0 init=/usr/lib/systemd/systemd no_timer_check noreplace-smp system
d.default_standard_error=journal+console systemd.default_standard_output=journal+console tsc=reliable MACHINEID=c6f15a6f-d37b-468b-9ef1-902ea984ec74 quiet root=/dev/root rw rootflags=rw,trans=virtio,version=9p2000.L rootfstype=9p
[    0.000000] KERNEL supported cpus:
[    0.000000]   Intel GenuineIntel
[    0.000000] CPU: vendor_id 'LKVMLKVMLKVM' unknown, using generic init.
[    0.000000] CPU: Your system may be unstable.
[    0.000000] e820: BIOS-provided physical RAM map:
[    0.000000] BIOS-e820: [mem 0x0000000000000000-0x000000000009fbff] usable
[    0.000000] BIOS-e820: [mem 0x000000000009fc00-0x000000000009ffff] reserved
[    0.000000] BIOS-e820: [mem 0x00000000000f0000-0x00000000000ffffe] reserved
[    0.000000] BIOS-e820: [mem 0x0000000000100000-0x0000000007ffffff] usable
[    0.000000] bootconsole [earlyser0] enabled


[    0.245978] ace-validator[74]: prestart OK
[    5.243048] ace-validator[75]: sidekick FAIL
[    5.243339] ace-validator[75]: ==> error touching file "/db/sidekick": open /db/sidekick: no such file or directory
[    5.243523] ace-validator[75]: ==> timed out waiting for /db/main
[    5.260207] ace-validator[81]: main FAIL
[    5.260496] ace-validator[81]: ==> error calling statfs on "/db": no such file or directory
[    5.260733] ace-validator[81]: ==> AC_APP_NAME not set appropriately (need "ace-validator-main", got "coreos.com/ace-validator-main")
[    5.260964] ace-validator[81]: ==> timed out waiting for /db/sidekick
[    5.261213] ace-validator[81]: ==> error touching file "/db/main": open /db/main: no such file or directory
[    5.283971] ace-validator[90]: poststop OK
[    5.354654] reboot: Restarting system with command '<rkt: obligatory restart>'
```

krnowak commented on stage1/init/mount.go:
```diff
@@ -0,0 +1,186 @@
+// file with functions that allow for creating a mount units for
+// managing inner and external volumes in stage1
+// mount units from systemd require to have /usr/bin/mount
+// TODO: it's would better to just use host@.mount empty@.mount systemd templates, but
```
Does this technique requires systemd 222 on host or in stage1? If in the latter only then please consider implementing it - we bumped systemd version in coreos to 222 recently (#1077).

krnowak commented on stage1/init/mount.go:
```diff
@@ -0,0 +1,186 @@
+		opts = append(opts, unit.NewUnitOption("Install", "RequiredBy", requiredBy))
+	}
+
+	unitsPath := filepath.Join(root, "/usr/lib/systemd/system")
```
Please use `unitsDir` from path.go instead of hardcoding `/usr/lib/systemd/system`.

krnowak commented on stage1/init/mount.go:
```diff
@@ -0,0 +1,186 @@
+	unitsPath := filepath.Join(root, "/usr/lib/systemd/system")
+	unitName := unit.UnitNamePathEscape(where + ".mount")
+
+	file, err := os.OpenFile(filepath.Join(unitsPath, unitName), os.O_WRONLY|os.O_CREATE, 0644)
```
Please consider using `ioutil.WriteFile()` instead of `osOpenFile()` and `io.Copy()`.

krnowak commented on stage1/init/mount.go:
```diff
@@ -0,0 +1,186 @@
+
+// appToSystemdMountUnits prepare bind mount unit for empty or host kind mouting
+// between stage1 rootfs and chrooted filesystem for application
+// func (p *Pod) appToSystemdMountUnits(ra *schema.RuntimeApp) error {
```
Please remove this line of comment.

krnowak commented on stage1/init/mount.go:
```diff
@@ -0,0 +1,186 @@
+
+		// make sure that "what" exists (created and mounted for pod)
+		log.Printf("checking required source path: %q", whatFullPath)
+		if _, err := os.Stat(whatFullPath); os.IsNotExist(err) {
```
I suppose that we should bail out here on any error from `os.Stat()`, no?

krnowak commented on stage1/init/mount.go:
```diff
@@ -0,0 +1,186 @@
+
+		// optionaly prepare app direcotry
+		log.Printf("optionally preapring destination path: %q", whereFullPath)
+		if _, err := os.Stat(whereFullPath); os.IsNotExist(err) {
```
What about scenario where `os.Stat()` returns some other error?

krnowak commented on stage1/init/pod.go:
```diff
@@ -214,6 +214,11 @@ func (p *Pod) appToSystemd(ra *schema.RuntimeApp, interactive bool) error {
 		unit.NewUnitOption("Service", "Group", "0"),
 	}
 
+	flavor, _, err := p.getFlavor()
```
`flavor` variable is used 100 lines lower, so please put it there.

ppalucki commented:
changes/updates:
- passing flavor down to functions (don't read it over and over again)
- removed os.Stat - we just rely on MkdirAll to return en error
- removed unused err, better error handling (stacking messages)
- better comments (no typos, easier to understand)
- general code cleaning (private functions, better names)
- removed dead code
- removed potential race condition between services & mounting (ordering for systemd units)
- using unitsDir
- added simple unittest 
- WriteFile instead of (OpenFile&Copy)

@krnowak: thx for review, good points, but I have more general question: what you think, about renaming `mount.go` to `kvm_mount.go` to indicate that this file (and functions) in only kvm flavor specific ?

(or even putting that in nested package stage1/init/kmv/ it will helps maintain a code separation when networking kvm specific code arrives)

alban commented on stage1/init/mount.go:
```diff
@@ -0,0 +1,199 @@
+// --volume=bar,kind=host,source=/host/bar
+// in stage1/rootfs rkt creates two folders (it is rootfs of guest)
+// 1. /mnt/foo - which is mounted through 9p host thanks to lkvm --9p=/host/bar,bar flag shared to quest
+// 2. /mnt/bar - is created as empty directory in guest
```
Above, `foo` is for `empty` and `bar` is for `host`. There is a mismatch.

I suggest calling the volumes `hostdata` and `temporary` in the examples to avoid confusion.

krnowak commented on stage1/init/mount.go:
```diff
@@ -0,0 +1,199 @@
+		}
+
+		// optionally prepare app directory
+		log.Printf("optionally preapring destination path: %q", whereFullPath)
```
Typo in `preapring`.

krnowak commented on stage1/init/mount.go:
```diff
@@ -0,0 +1,199 @@
+	return nil
+}
+
+// PodToKvmArgs renders a prepared Pod
```
Wrong function name in docs.

krnowak commented on stage1/init/mount.go:
```diff
@@ -0,0 +1,199 @@
+// Copyright 2014 The rkt Authors
```
It is 2015 now. :)

krnowak commented on stage1/init/mount_test.go:
```diff
@@ -0,0 +1,65 @@
+// Copyright 2014 The rkt Authors
```
2015.

krnowak commented:
@ppalucki: Small nitpicks, but LFAD otherwise (looks fine and dandy).

About moving the code to kvm subpackage - yes, please put mount.go and mount_test.go in kvm subpackage. The stage1/init package needs that splitting. Thanks for suggesting.

krnowak commented:
@ppalucki: I see that you addressed all my issues. Are you going to move the `mount{,_test}.go` files to kvm subpackage or do you prefer to do it in separate PR?

ppalucki commented:
@krnowak: give me one more day, I will move kvm code into separate package - it would benefit it further PRs

krnowak commented:
Alright.

iaguis commented:
About the mount problem, see https://github.com/appc/spec/pull/467

ppalucki commented:
update:
- moved all *kvm* code to separate package (the same done for [networking support](#1219)), but had to copy-paste one function from path.go [here](https://github.com/coreos/rkt/pull/1176/files#diff-907b987e11100433965d99c87a6cbd76R50)  (TODOed for later)

@krnowak: still needs some functional tests, but I'm not going to refactor this anymore - or if I can do this in new PR

jonboulle commented on stage1/usr_from_coreos/manifest.d/systemd:
```diff
@@ -1,6 +1,7 @@
 bin/coredumpctl
 bin/journalctl
 bin/systemctl
+bin/mount
```
Sort me, please

alban replied:
Is this change needed? The `.mount` files are created only in the `kvm` flavor, so it should not be needed here in the `coreos` flavor? If it is actually needed by systemd, maybe `umount` will be needed as well to stop the pod cleanly.

krnowak replied:
Well, `kvm` flavor is based on `coreos` flavor. Currently there is no way to specify an additional manifest.

alban replied:
Oh, I see... `usr_from_kvm.mk` includes `../usr_from_coreos/usr_from_coreos.mk`.

jonboulle commented on stage1/init/kvm/mount.go:
```diff
@@ -0,0 +1,208 @@
+			return fmt.Errorf("failed to prepare dir for mountPoint %v: %v", mountPoint.Name, err)
+		}
+
+		// install new mount unit for bind mount /mnt/volumenName -> /opt/stage2/{app-id}/rootfs/{{mountPoint.Path}}
```
volumeName

jonboulle commented on stage1/init/kvm/mount.go:
```diff
@@ -0,0 +1,208 @@
+	return appName.String() + ".service"
+}
+
+// installNewMountUnit utility function creates and installs new mount unit
```
s/utility function//

jonboulle commented on stage1/init/kvm/mount.go:
```diff
@@ -0,0 +1,208 @@
+	return nil
+}
+
+// VolumesToKvmDiskArgs renders a prepared Pod
```
line wrapping is overly aggressive here

jonboulle commented on stage1/init/kvm/mount.go:
```diff
@@ -0,0 +1,208 @@
+			unitsDir,
+		)
+		if err != nil {
+			return fmt.Errorf("cannot install new mount unit error for app %q: %v", appName.String(), err)
```
s/error //

alban commented on stage1/init/kvm/mount.go:
```diff
@@ -0,0 +1,208 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// mount.go provides functions for creating a mount units for
```
`a mount units`: singular or plural?

ppalucki commented:
update:
- comments fixed and aligned to 80 column width
- bin/mount reordered and becouse I didn't want to hack usr-from-coreos flavor to include bin/mount or rewrite copying all the extraction mechanism to  kvm flavor - let's stay with bin/mount added to coreos flavor, ok?

ps. I need to change my workflow and stop rebasing/squashing commits (then I don't need to "write" updates with every push :smile: )

krnowak commented:
LFAD.
